### DISCUSSION TITLE
How to pretty-print an OFX tree?

### DISCUSSION THREAD

**Discussion by jeffstearns** [csingley/ofxtools#167] [Q&A]
What's the best way to pretty-print an OFX tree?

I'm trying to pretty-print the result from parser.convert() for debug purposes.  I want to walk the tree, printing each node in a human-readable format.

I've combed through various projects using ofxtools, but haven't been able to find code that does this.

I'm having moderate success using the pprint module.  It's possible, but I'm using quite a bit of monkey-patching that depends on private implementation details.

Looking at the `__repr__()` implementation in https://github.com/csingley/ofxtools/blob/master/ofxtools/models/base.py suggests that there should be some elegant solution for walking an OFX tree and dumping the nodes.  But I can't quite get my head around the best way to do that.

Any suggestions?  The output just needs to be human-readable.  It doesn't need to be in a format that can be fed into any other program.

**Comment by csingley** [comment]
Are you looking for ofxtools.utils.indent() ?

https://github.com/csingley/ofxtools/blob/655549a357d9e5109460ab4687fa9e304694e17c/ofxtools/utils.py#L101

**Comment by jeffstearns** [comment]
Hi, thanks for your interest.

I'm trying to write a function that does a tree walk.  I pass in an element of an OFX tree, and the function walks the entire subtree, recursively visiting each sub-element.

The code I've prototyped is fragile and dependent on private implementation details.  I'd much prefer to stick to interfaces that are more public.

Looking through the ofxtools source, I find code that looks suggestive, but I typically run into trouble using it to discover the exact set of child nodes.  It's the dynamic nature of ET.Elements that confuses me.  (I think my brain is addled by metaprogramming.)

Is there a function that returns the children of a given node?  Something like
`def children(elem: ET.Element) -> iterator over all nodes below elem`

**Reply by csingley** [reply]
All `OFXTree` does is subclass `ElementTree.ElementTree` and override the `parse()` method.  Everything else is the same.  The whole point is that you can keep using the well-known [`ElementTree` interface](https://docs.python.org/3/library/xml.etree.elementtree.html#xml.etree.ElementTree.ElementTree).

Maybe get the root node; that's an `ElementTree.Element` and you can use that [well-known interface](https://docs.python.org/3/library/xml.etree.elementtree.html#element-objects).

If you don't know how to use ElementTree, you're gonna have a bad time manipulating `OFXTree` data structures.  I'd suggest to go brush up on that first & then come back to play with `ofxtools`.
